### PR TITLE
Implement Authentication

### PR DESCRIPTION
**Pull request by George-Miao**
Implementing #53

### CODE REVIEW

File: auth/src/model.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+/// A partial map whose domain are central components and co-domain are read-only and read-write.
+#[must_use]
+#[non_exhaustive]
+#[derive(Clone, Debug, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct PermissionSet {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub api: Option<Permission>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub method: Option<Permission>,
```
Comment (PhotonQuantum) [inline auth/src/model.rs:24]: mq?

File: auth/src/model.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    /// Empty permission set, does not have access to any of the components.
+    pub const EMPTY: Self = Self::empty();
+
+    /// Full permission set, has access to all of the components.
+    pub const FULL: Self = Self::full();
+
+    pub(crate) const fn empty() -> Self {
+        Self {
+            api: None,
+            method: None,
```
Comment (PhotonQuantum) [inline auth/src/model.rs:39]: ditto

File: auth/src/model.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            coordinator: None,
+        }
+    }
+
+    pub(crate) const fn full() -> Self {
+        Self {
+            api: Some(Permission::ReadWrite),
+            method: Some(Permission::ReadWrite),
```
Comment (PhotonQuantum) [inline auth/src/model.rs:47]: ditto

File: auth/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    /// Try insert a new record.
+    ///
+    /// Return wether the record is inserted.
+    /// If one record with same username exists, this will leave it intact.
+    pub async fn new_record(
+        &self,
+        username: impl Into<String>,
+        password: &[u8],
```
Comment (PhotonQuantum) [inline auth/src/lib.rs:70]: impl AsRef<[u8]>?

File: auth/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        Ok(res.upserted_id.is_some())
+    }
+
+    /// Look up permission of a user by username and password.
+    ///
+    /// When the username and password combination are invalid, this will return [`PermissionSet::EMPTY`].
+    pub async fn look_up(
+        &self,
+        username: impl AsRef<str>,
+        password: &[u8],
```
Comment (PhotonQuantum) [inline auth/src/lib.rs:102]: ditto

Comment (George-Miao) [reply auth/src/lib.rs]: Updated

File: auth/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+argon2    = { version = "0.4.0", features = ["std"] }
+serde     = { version = "1.0.136", features = ["derive"] }
+mongodb   = { version = "2.2.0-beta", features = ["bson-uuid-0_8"] }
+
+[dev-dependencies]
+futures = "0.3.21"
+serde_json = "1.0.79"
+tokio      = { version = "1.17.0", features = ["macros"] }
+
+[features]
```
Comment (PhotonQuantum) [inline auth/Cargo.toml:20]: Please reformat this file

File: auth/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+mod_use   = "0.2.0"
+thiserror = "1.0.30"
+argon2    = { version = "0.4.0", features = ["std"] }
+serde     = { version = "1.0.136", features = ["derive"] }
+mongodb   = { version = "2.2.0-beta", features = ["bson-uuid-0_8"] }
```
Comment (PhotonQuantum) [inline auth/Cargo.toml:13]: sort dependencies

Comment (PhotonQuantum) [reply auth/Cargo.toml:13]: and should we lock patch or minor?

Comment (George-Miao) [reply auth/Cargo.toml:13]: > and should we lock patch or minor?

This is equivalent to ^ in semver, which allows patch update.

Comment (George-Miao) [reply auth/Cargo.toml:13]: > sort dependencies

No. This sorting is intended. Sort according to length from short to long.

File: auth/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                },
+                doc! {
+                 "$setOnInsert": doc
+                },
+                UpdateOptions::builder().upsert(true).build(),
+            )
+            .await?;
+
+        Ok(res.upserted_id.is_some())
+    }
```
Comment (PhotonQuantum) [inline auth/src/lib.rs:94]: This function can only attach permissions to users during creation. So how do we modify permissions of a user?

**Pull request merged**